api/resources_portal/models/notification.py

```diff
@@ -0,0 +1,92 @@
+import logging
+
+from django.db import models
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+
+from computedfields.models import ComputedFieldsModel, computed
+
+from resources_portal.models.material import Material
+from resources_portal.models.organization import Organization
+from resources_portal.models.user import User
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+logger.addHandler(logging.StreamHandler())
+
+
+class Notification(ComputedFieldsModel):
+    class Meta:
+        db_table = "notifications"
+        get_latest_by = "created_at"
+
+    NOTIFICATION_TYPES = (
+        ("ORG_REQUEST_CREATED", "ORG_REQUEST_CREATED"),
+        ("ORG_INVITE_CREATED", "ORG_INVITE_CREATED"),
+        ("ORG_INVITE_ACCEPTED", "ORG_INVITE_ACCEPTED"),
+        ("ORG_REQUEST_ACCEPTED", "ORG_REQUEST_ACCEPTED"),
+        ("ORG_INVITE_REJECTED", "ORG_INVITE_REJECTED"),
+        ("ORG_REQUEST_REJECTED", "ORG_REQUEST_REJECTED"),
+        ("ORG_INVITE_INVALID", "ORG_INVITE_INVALID"),
+        ("ORG_REQUEST_INVALID", "ORG_REQUEST_INVALID"),
+        ("MTA_UPLOADED", "MTA_UPLOADED"),
+        ("APPROVE_REQUESTS_PERM_GRANTED", "APPROVE_REQUESTS_PERM_GRANTED"),
+        ("TRANSFER_REQUESTED", "TRANSFER_REQUESTED"),
+    )
+
+    objects = models.Manager()
+
+    created_at = models.DateTimeField(auto_now_add=True)
+
+    notification_type = models.CharField(max_length=32, choices=NOTIFICATION_TYPES)
+    notified_user = models.ForeignKey(
+        User, blank=False, null=False, on_delete=models.CASCADE, related_name="notifications"
+    )
+    associated_user = models.ForeignKey(
+        User,
+        blank=False,
+        null=True,
+        on_delete=models.CASCADE,
+        related_name="associated_notifications",
+    )
+    associated_organization = models.ForeignKey(
+        Organization, blank=False, null=True, on_delete=models.CASCADE
+    )
+    associated_material = models.ForeignKey(
+        Material, blank=False, null=True, on_delete=models.CASCADE
+    )
+
+    @computed(models.TextField(null=False, blank=False))
+    def message(self):
```

Comment:
This block is ugly as sin, but I couldn't figure out a better way to do it. Suggestions would be very welcome.

Reply:
https://www.python.org/dev/peps/pep-3103/

Reply:
or something like this 
```python

alert_messages = {
"ORG_REQUEST_CREATED" : "{} is requesting to to join {}.".format(self.associated_user.username, self.associated_organization.name)
...
}

try:
    message = alert_messages[self.response_type]
except ValueError:
    raise ValueError(f'"{self.notification_type}" is not a valid notification type')
```

Reply:
I like the `alert_messages` dict idea!

Reply:
Much better- thank you.

Reply:
I'm not sure if @kurtwheeler would go for it but i think this is slick
```python 
>>> class Test(object):
...     def __init__(self, a, b):
...         self.a = a
...         self.b = b
...
>>> a = "one"
>>> b = Test("two", "three")
>>> c = Test(a, b)
>>> demo = "{a} {b.b}"
>>> demo.format(**c.__dict__)
'one three'
```

Reply:
though either way you will want to determine your attributes to apply to the string before hand so you don't get an exception when creating the templates